# Employee Performance Prediction

## Machine Learning Model Development

### Objective

The objective of this notebook is to build machine learning models that predict employee performance ratings based on employee-related features.

### Workflow

1. Load Processed Dataset
2. Feature Selection
3. Feature Encoding
4. Train-Test Split
5. Model Training
6. Model Evaluation
7. Feature Importance Analysis

In [1]:
import pandas as pd
import numpy as np 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (accuracy_score ,classification_report,confusion_matrix)

## Loading Processed Dataset

In [2]:
df=pd.read_csv(r"C:\Users\nimba\Desktop\INX_Employee_Performance_Project\data\processed\employee_performance_processed.csv")
df.head()

,EmpNumber,Age,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,DistanceFromHome,EmpEducationLevel,...,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,Attrition,PerformanceRating
0,E1001000,32,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,10,3,...,4,10,2,2,10,7,0,8,No,3
1,E1001006,47,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,14,4,...,4,20,2,3,7,7,1,7,No,3
2,E1001007,40,Male,Life Sciences,Married,Sales,Sales Executive,Travel_Frequently,5,4,...,3,20,2,3,18,13,1,12,No,4
3,E1001009,41,Male,Human Resources,Divorced,Human Resources,Manager,Travel_Rarely,10,4,...,2,23,2,2,21,6,12,6,No,3
4,E1001010,60,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,16,4,...,4,10,1,3,2,2,2,2,No,3


In [3]:
df.shape

(1200, 28)

In [4]:
df.select_dtypes(include="object").columns

Index(['EmpNumber', 'Gender', 'EducationBackground', 'MaritalStatus',
       'EmpDepartment', 'EmpJobRole', 'BusinessTravelFrequency', 'OverTime',
       'Attrition'],
      dtype='object')

## Feature Encoding

Machine learning algorithms require numerical input. Therefore, categorical variables are converted into numerical values using Label Encoding.

In [5]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
categorical_columns=df.select_dtypes(include="object").columns
for col in categorical_columns:
    df[col]=le.fit_transform(df[col])
df.head()    

,EmpNumber,Age,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,DistanceFromHome,EmpEducationLevel,...,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,Attrition,PerformanceRating
0,0,32,1,2,2,5,13,2,10,3,...,4,10,2,2,10,7,0,8,0,3
1,1,47,1,2,2,5,13,2,14,4,...,4,20,2,3,7,7,1,7,0,3
2,2,40,1,1,1,5,13,1,5,4,...,3,20,2,3,18,13,1,12,0,4
3,3,41,1,0,0,3,8,2,10,4,...,2,23,2,2,21,6,12,6,0,3
4,4,60,1,2,2,5,13,2,16,4,...,4,10,1,3,2,2,2,2,0,3


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 28 columns):
 #   Column                        Non-Null Count  Dtype
---  ------                        --------------  -----
 0   EmpNumber                     1200 non-null   int32
 1   Age                           1200 non-null   int64
 2   Gender                        1200 non-null   int32
 3   EducationBackground           1200 non-null   int32
 4   MaritalStatus                 1200 non-null   int32
 5   EmpDepartment                 1200 non-null   int32
 6   EmpJobRole                    1200 non-null   int32
 7   BusinessTravelFrequency       1200 non-null   int32
 8   DistanceFromHome              1200 non-null   int64
 9   EmpEducationLevel             1200 non-null   int64
 10  EmpEnvironmentSatisfaction    1200 non-null   int64
 11  EmpHourlyRate                 1200 non-null   int64
 12  EmpJobInvolvement             1200 non-null   int64
 13  EmpJobLevel                   120

## Feature and Target Separation

The target variable is separated from the predictor variables before model training.

In [7]:
x=df.drop("PerformanceRating",axis=1)
y=df["PerformanceRating"]
print("Feature shape:",x.shape)
print("Target shape:",y.shape)

Feature shape: (1200, 27)
Target shape: (1200,)


## Train-Test Split

The dataset is divided into training and testing sets. The training set is used to build the model, while the testing set is used to evaluate model performance.

In [8]:
X = df.drop("PerformanceRating", axis=1)

y = df["PerformanceRating"]

print("Feature Shape:", X.shape)
print("Target Shape:", y.shape)

Feature Shape: (1200, 27)
Target Shape: (1200,)


In [9]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=42)

print("x_train:", x_train.shape)
print("x_test:", x_test.shape)

x_train: (960, 27)
x_test: (240, 27)


In [10]:
print(df.columns.tolist())

['EmpNumber', 'Age', 'Gender', 'EducationBackground', 'MaritalStatus', 'EmpDepartment', 'EmpJobRole', 'BusinessTravelFrequency', 'DistanceFromHome', 'EmpEducationLevel', 'EmpEnvironmentSatisfaction', 'EmpHourlyRate', 'EmpJobInvolvement', 'EmpJobLevel', 'EmpJobSatisfaction', 'NumCompaniesWorked', 'OverTime', 'EmpLastSalaryHikePercent', 'EmpRelationshipSatisfaction', 'TotalWorkExperienceInYears', 'TrainingTimesLastYear', 'EmpWorkLifeBalance', 'ExperienceYearsAtThisCompany', 'ExperienceYearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager', 'Attrition', 'PerformanceRating']


In [11]:
print("EmpNumber" in df.columns)

True


In [12]:
print(df.shape)
print(X.shape)

(1200, 28)
(1200, 27)


## Decision Tree Model

A Decision Tree Classifier is trained to predict employee performance ratings using employee-related features.

In [13]:
from sklearn.tree import DecisionTreeClassifier
dt_model=DecisionTreeClassifier(random_state=42)
dt_model.fit(x_train,y_train)
y_pred_dt=dt_model.predict(x_test)

In [14]:
from sklearn.metrics import accuracy_score
accuracy_dt=accuracy_score(y_test,y_pred_dt)
print("Decision Tree Accuracy:",accuracy_dt)

Decision Tree Accuracy: 0.8916666666666667


In [15]:
from sklearn.metrics import classification_report

print(classification_report(y_test,y_pred_dt))

              precision    recall  f1-score   support

           2       0.79      0.93      0.86        29
           3       0.95      0.90      0.93       184
           4       0.66      0.78      0.71        27

    accuracy                           0.89       240
   macro avg       0.80      0.87      0.83       240
weighted avg       0.90      0.89      0.89       240



## Decision Tree Model

A Decision Tree Classifier is used to predict employee performance ratings based on employee-related attributes.

### Why Decision Tree?

Decision Trees are easy to interpret and help identify the factors influencing employee performance. The model works by creating a series of decision rules based on the input features and classifies employees into different performance rating categories.

### Advantages

- Easy to understand and visualize
- Handles both numerical and categorical features
- Captures complex relationships between variables
- Provides insights into feature importance
- Suitable for employee performance prediction problems

The model is trained using the employee dataset and evaluated using accuracy, precision, recall, and F1-score metrics.

## Decision Tree Model Results

The Decision Tree Classifier achieved an accuracy of 89.17% on the test dataset.

### Performance Summary

- Rating 2 employees were identified with high recall (93%).
- Rating 3 employees were predicted with excellent precision (95%) and recall (90%).
- Rating 4 employees were predicted with moderate performance due to the smaller number of observations in this category.

Overall, the model demonstrates strong predictive capability and serves as a reliable baseline model for employee performance prediction.

## Random Forest Model

Random Forest is an ensemble machine learning algorithm that combines multiple Decision Trees to improve prediction accuracy and reduce overfitting.

### Why Random Forest?

- More accurate than a single Decision Tree
- Reduces overfitting
- Produces stable predictions
- Provides feature importance scores
- Suitable for classification problems

The Random Forest model is trained to predict employee performance ratings and identify the most important factors affecting employee performance.

In [24]:
x.columns

Index(['EmpNumber', 'Age', 'Gender', 'EducationBackground', 'MaritalStatus',
       'EmpDepartment', 'EmpJobRole', 'BusinessTravelFrequency',
       'DistanceFromHome', 'EmpEducationLevel', 'EmpEnvironmentSatisfaction',
       'EmpHourlyRate', 'EmpJobInvolvement', 'EmpJobLevel',
       'EmpJobSatisfaction', 'NumCompaniesWorked', 'OverTime',
       'EmpLastSalaryHikePercent', 'EmpRelationshipSatisfaction',
       'TotalWorkExperienceInYears', 'TrainingTimesLastYear',
       'EmpWorkLifeBalance', 'ExperienceYearsAtThisCompany',
       'ExperienceYearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager', 'Attrition'],
      dtype='object')

In [26]:
x= df.drop(["EmpNumber","PerformanceRating"], axis=1)

In [29]:
from sklearn.metrics import accuracy_score
accuracy_rf=accuracy_score(y_test,y_pred_rf)
print("Random Forest Accuracy:",accuracy_rf)

Random Forest Accuracy: 0.95


In [30]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred_rf))

              precision    recall  f1-score   support

           2       0.93      0.90      0.91        29
           3       0.95      0.99      0.97       184
           4       1.00      0.74      0.85        27

    accuracy                           0.95       240
   macro avg       0.96      0.88      0.91       240
weighted avg       0.95      0.95      0.95       240



## Random Forest Model Results

The Random Forest Classifier achieved an accuracy of 95% on the test dataset, outperforming the Decision Tree model.

### Performance Summary

- Rating 2 employees were predicted with 93% precision and 90% recall.
- Rating 3 employees were predicted with 95% precision and 99% recall.
- Rating 4 employees were predicted with 100% precision and 74% recall.

The Random Forest model demonstrated excellent predictive performance and was selected as the final model for employee performance prediction.

In [34]:
print(X.shape)

(1200, 26)


In [35]:
print(len(X.columns))

26


In [36]:
print(len(rf_model.feature_importances_))

27


In [37]:
X = df.drop("PerformanceRating", axis=1)

y = df["PerformanceRating"]

print(X.shape)

(1200, 27)


In [40]:
df = df.drop("EmpNumber", axis=1)

print(df.columns.tolist())

['Age', 'Gender', 'EducationBackground', 'MaritalStatus', 'EmpDepartment', 'EmpJobRole', 'BusinessTravelFrequency', 'DistanceFromHome', 'EmpEducationLevel', 'EmpEnvironmentSatisfaction', 'EmpHourlyRate', 'EmpJobInvolvement', 'EmpJobLevel', 'EmpJobSatisfaction', 'NumCompaniesWorked', 'OverTime', 'EmpLastSalaryHikePercent', 'EmpRelationshipSatisfaction', 'TotalWorkExperienceInYears', 'TrainingTimesLastYear', 'EmpWorkLifeBalance', 'ExperienceYearsAtThisCompany', 'ExperienceYearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager', 'Attrition', 'PerformanceRating']


In [41]:
print(df.shape)

(1200, 27)


In [42]:
X = df.drop("PerformanceRating", axis=1)

y = df["PerformanceRating"]

print(X.shape)

(1200, 26)


In [43]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,  y,  test_size=0.2,random_state=42)

In [44]:
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier( n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)

In [45]:
print(len(X.columns))
print(len(rf_model.feature_importances_))

26
26


In [46]:
feature_importance = pd.DataFrame({  "Feature": X.columns,"Importance": rf_model.feature_importances_})
feature_importance = feature_importance.sort_values( by="Importance", ascending=False)
feature_importance.head(10)

,Feature,Importance
9,EmpEnvironmentSatisfaction,0.219013
16,EmpLastSalaryHikePercent,0.212347
23,YearsSinceLastPromotion,0.090565
22,ExperienceYearsInCurrentRole,0.043961
4,EmpDepartment,0.039529
5,EmpJobRole,0.037145
10,EmpHourlyRate,0.035377
0,Age,0.032529
21,ExperienceYearsAtThisCompany,0.031829
7,DistanceFromHome,0.027037


## Feature Importance Analysis

Feature importance analysis was performed using the Random Forest Classifier to identify the key factors influencing employee performance ratings.

### Top Factors Affecting Employee Performance

1. Employee Environment Satisfaction
2. Last Salary Hike Percentage
3. Years Since Last Promotion
4. Experience in Current Role
5. Employee Department
6. Employee Job Role
7. Employee Hourly Rate
8. Employee Age
9. Experience at Current Company
10. Distance From Home

The analysis indicates that work environment and compensation-related factors have the strongest influence on employee performance.

## Business Insights

The analysis revealed several important factors affecting employee performance.

### Key Findings

- Employee Environment Satisfaction is the most influential factor affecting performance ratings.
- Employees receiving higher salary hikes tend to demonstrate better performance.
- Promotion history significantly impacts employee performance.
- Employees with greater experience in their current role generally perform better.
- Job role and department influence employee productivity and performance levels.
- Compensation and workplace satisfaction are stronger predictors than demographic characteristics.

### Recommendations

- Improve workplace environment and employee satisfaction programs.
- Introduce fair and performance-based salary increment policies.
- Create transparent promotion and career growth opportunities.
- Provide role-specific training and skill development programs.
- Focus on employee engagement and retention initiatives.

## Model Comparison

Two machine learning models were developed and evaluated for predicting employee performance ratings.

| Model | Accuracy |
|---------|---------|
| Decision Tree Classifier | 89.17% |
| Random Forest Classifier | 95.00% |

The Random Forest model achieved the highest accuracy and was selected as the final model for performance prediction.

# Conclusion

The objective of predicting employee performance ratings was successfully achieved using machine learning techniques.

A comprehensive exploratory data analysis was conducted to understand the factors influencing employee performance. Multiple business insights were identified through statistical analysis and visualization.

Two classification models were developed:

1. Decision Tree Classifier
2. Random Forest Classifier

The Random Forest Classifier achieved the best performance with an accuracy of 95%, outperforming the Decision Tree model.

Feature importance analysis revealed that Employee Environment Satisfaction, Last Salary Hike Percentage, and Years Since Last Promotion are the most significant factors affecting employee performance.

The developed model can assist HR departments in identifying performance drivers, improving employee satisfaction, and making data-driven workforce management decisions.